File: May-LeetCoding-Challenge/04-Number-Complement/Number-Complement.cpp
class Solution {
public:
    int findComplement(int num) {
        long i = 1;
        while (i <= num) i <<= 1;
        return (i-1) ^ num;
    }
};

class Solution2 {
private:
    inline int fill_bits(int v) {
        v |= v >> 1;
        v |= v >> 2;
        v |= v >> 4;
        v |= v >> 8;
        v |= v >> 16;
        return v;
    }
public:

    int findComplement(int num) {
        return fill_bits(num + (num==0) ) & ~num;
    }
};

File: May-LeetCoding-Challenge/04-Number-Complement/Number-Complement.scala
object Solution1 {
  def findComplement(num: Int): Int = {
    num
      .toBinaryString
      .map(c => if (c == '1') '0' else '1')
      .dropWhile(_ != '1')
      .reverse
      .zipWithIndex
      .map(t => t._1.toString.toInt * math.pow(2,t._2))
      .sum
      .toInt

  }
}

//with java function Integer.parseInt
object Solution2 {
  def findComplement(num: Int): Int = Integer.parseInt(num.toBinaryString.map(c => if (c == '1') '0' else '1'), 2)
}

//learn from others  ^^
object Solution3 {
  def findComplement(num: Int): Int = {
    var mask = 1
    while (mask < num) {
      mask = (mask << 1) + 1
    }
    num ^ mask
  }
}